Subscription-based premium themes: Support multiple premium theme products

DavidRothstein opened the pull request:
#### Proposed Changes

Currently Calypso assumes all premium themes have the same product slug.  But to support having subscription-based themes with different products for each theme (an implementation we are currently investigating) it is necessary to make Calypso check the product type rather than the product slug to determine if a purchase is a premium theme.

This pull request accomplishes that. Props to @mreishus for part of the code here (from https://github.com/Automattic/wp-calypso/pull/66319).

Related to: paYKcK-1Qr-p2

#### Testing Instructions

The most important testing is to buy a premium theme via `/themes/premium/[YOURSITE]` as a one-time purchase and make sure the entire experience is unchanged by this pull request:
1. Buy the theme and make sure it can be activated during the purchase process.
2. Go back to `/themes/premium/[YOURSITE]` (with the purchased theme either set as active or not) and make sure it displays correctly.
3. Go to `/themes/premium/[YOURSITE]?flags=signup/seller-upgrade-modal` (per https://github.com/Automattic/wp-calypso/pull/66285#issuecomment-1206744656) and make sure it looks correct there too. Note that it looks like the theme is being misidentified as an annual subscription on this page (even though it's a one-time purchase) but that is a preexisting issue.
4. Go to `/setup/designSetup?siteSlug=[YOURSITE]&flags=signup/design-picker-unified` (per https://github.com/Automattic/wp-calypso/pull/66319) and make sure it looks correct there too. As with the above, it looks like the theme is being misidentified as an annual subscription on this page (even though it's a one-time purchase) but that is a preexisting issue.
5. Cancel/refund the theme and make sure that works correctly also.

You can also test with D85381-code on the server and follow the instructions there to test purchasing an annual subscription. Following similar testing steps as above, the theme should display in a reasonable way everywhere in the user interface.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

DavidRothstein committed:
Subscription-based premium themes: Support multiple premium theme products

DavidRothstein commented on packages/calypso-products/src/camel-or-snake-product-type.ts:
```diff
@@ -0,0 +1,7 @@
+export function camelOrSnakeProductType(
+	product: WithCamelCaseProductType | WithSnakeCaseProductType
+): string {
+	return 'product_type' in product ? product.product_type : product.productType;
```
I'm not sure if it's actually necessary to support both, but the existing code for the product slug does the same thing, so I did it here too.

sirbrillig replied:
For product slugs this is because we have so many different variations of what a "product" object is (cart product, receipt product, products endpoint product, etc.) and they have different slug formatting. While there's nothing really wrong with using the same approach here, if this is only going to be used for one type of product object, I'd be inclined to stick to just that one type of slug. There's no need to support a property that doesn't exist anywhere.

mreishus replied:
There's something odd going on in calypso, where [snake case responses are mapped to camel case](https://github.com/Automattic/wp-calypso/blob/e7326458f8f0369c9f6c8a132c32b92b27265a08/client/lib/purchases/assembler.js#L77-L79), and [test data uses snake case](https://github.com/Automattic/wp-calypso/blob/c17f00d14ca19be8e25f485dca3b892b1a3d6058/client/state/themes/test/selectors.js#L2241), but my selectors checking those fields in that data [in camel case](https://github.com/Automattic/wp-calypso/blob/c17f00d14ca19be8e25f485dca3b892b1a3d6058/client/state/themes/selectors/get-purchased-themes.js#L17) work.
Could this be to support both untransformed and transformed objects?

sirbrillig reviewed (commented):
This looks fine to me (I haven't run through the tests yet) but I do wonder about the semantics of `isTheme()`.

Previously, that function could have been used on any type of object that had a slug (eg: shopping cart product, purchases list product, receipt item product), but now it will only work for one specific type of object (purchases list product).

In practice this may be fine since the only places using it currently seem to pass it products list data. Also, this is precisely what TS is for, and should prevent passing it an object that is lacking the required property. Hopefully everywhere that calls it in the future uses TS. 🤞 

I made one suggestion to avoid premature optimization unless you think there's a good reason to support different types of slugs.

mreishus commented:
:+1: I've purchased both legacy premium theme (skivers) and a new annual subscription (baxter) one. Calypso recognizes both:

![2022-08-05_13-31](https://user-images.githubusercontent.com/937354/183139708-fd226a91-b515-4e98-82b1-b9f773fcf416.png)

It also does with `?flags=signup/seller-upgrade-modal`

![2022-08-05_13-35](https://user-images.githubusercontent.com/937354/183140240-fba2555c-3511-4c39-bb37-e40c0f4faabb.png)


Billing screens look good:

![2022-08-05_13-16](https://user-images.githubusercontent.com/937354/183139848-abb9dd34-9958-4d12-a6ee-fed8a25edbfd.png)
![2022-08-05_13-15](https://user-images.githubusercontent.com/937354/183139868-57729372-b527-4380-b4da-8b3fa00e58cd.png)

Purchase process worked fine (ignore sandbox warnings in preview modal)
![2022-08-05_13-12](https://user-images.githubusercontent.com/937354/183139968-c49c9a06-1fdf-4a0f-8690-6ba3beb5defc.png)
![2022-08-05_13-14](https://user-images.githubusercontent.com/937354/183140002-3a9962e9-f7ed-47cb-b19a-49a06d5f3aa2.png)

mreishus commented:
The unified design picker on (`/setup/designSetup` with `signup/design-picker-unified` flag on) is not noticing the new purchases:
![2022-08-05_13-39](https://user-images.githubusercontent.com/937354/183140926-30495449-1e77-4f69-b0df-5c84fc48bee3.png)
This is something I just worked on though, so **I will create a small PR that fixes** (or can be added to this PR).

DavidRothstein committed:
Review feedback: Consistently use productType, update getPurchasedThemes(), etc.

DavidRothstein commented:
Thanks for the reviews!

After checking all usages of `isTheme()` in the codebase, it looks like all of them pass in a purchase object which came from either `createPurchaseObject()` (which this diff adds the `productType` property to) or `createReceiptObject()` (which already had a `productType` property).  So I agree; we only need to support that one property.  I've simplified the pull request accordingly. Also, I renamed `isTheme()` to `isThemePurchase()` to hopefully be a bit more clear about what is expected to be passed in.

@mreishus Thanks so much for posting https://github.com/Automattic/wp-calypso/pull/66319! (I believe that new `getPurchasedThemes()` method was introduced right around the time I was working on this).  I've incorporated those changes here, and I think it works well.

I've added some updated testing instructions to the main body of the pull request, and I think this is hopefully ready to go in.

DavidRothstein committed:
Fix TypeScript errors

sirbrillig reviewed (approved):
Since the `product_type` field is present since D85691-code, these changes look good to me. I think they're semantic and seem to work well.
